internal/transactionsubmission/engine/signing/signature_service.go

```diff
@@ -0,0 +1,108 @@
+package signing
+
+import (
+	"fmt"
+
+	"golang.org/x/exp/slices"
+
+	"github.com/stellar/stellar-disbursement-platform-backend/db"
+	"github.com/stellar/stellar-disbursement-platform-backend/internal/transactionsubmission/engine/preconditions"
+	"github.com/stellar/stellar-disbursement-platform-backend/internal/transactionsubmission/utils"
+)
+
+type SignatureService struct {
+	ChAccountSigner   SignatureClient
+	DistAccountSigner SignatureClient
+	HostAccountSigner SignatureClient
+	DistributionAccountResolver
+}
+
+var _ DistributionAccountResolver = (*SignatureService)(nil)
+
+func (s *SignatureService) Validate() error {
+	if s.ChAccountSigner == nil {
+		return fmt.Errorf("channel account signer cannot be nil")
+	}
+
+	if s.DistAccountSigner == nil {
+		return fmt.Errorf("distribution account signer cannot be nil")
+	}
+
+	if s.HostAccountSigner == nil {
+		return fmt.Errorf("host account signer cannot be nil")
+	}
+
+	if s.ChAccountSigner.NetworkPassphrase() != s.DistAccountSigner.NetworkPassphrase() || s.DistAccountSigner.NetworkPassphrase() != s.HostAccountSigner.NetworkPassphrase() {
+		return fmt.Errorf("network passphrase of all signers should be the same")
+	}
+
+	if s.DistributionAccountResolver == nil {
+		return fmt.Errorf("distribution account resolver cannot be nil")
+	}
+
+	return nil
+}
+
+type SignatureServiceOptions struct {
+	// Shared:
+	NetworkPassphrase string
+
+	// DistributionAccount:
+	DistributionPrivateKey string
+	// DistributionAccountEnv:
+	DistributionSignerType SignatureClientType
+
+	// ChannelAccountDB:
+	DBConnectionPool     db.DBConnectionPool
+	EncryptionPassphrase string
+	LedgerNumberTracker  preconditions.LedgerNumberTracker
+	Encrypter            utils.PrivateKeyEncrypter
+}
+
+// NewSignatureService creates a new signature service instance, given the distribution signer type and the options.
+func NewSignatureService(opts SignatureServiceOptions) (SignatureService, error) {
```

Comment:
Could you explain the difference between the `SignatureClient` and `SignatureService`?

Reply:
The signature service needs to be able to sign for three account types,

- ChannelAccounts
- (per tenant) distribution accounts
- host account

Every one of these could have a different signer client, and that's by design, considering we want to treat development accounts with higher security than channel accounts, for instance.

In the future, if we decide to use a single signer for all account types, the structure of the SIgnatureService should be fairly easy to adapt into that approach.